nntrainer/dataset/raw_file_data_producer.cpp

```diff
@@ -0,0 +1,180 @@
+                          const std::vector<TensorDim> &label_dims) const {
+  auto size_accumulator = [](const unsigned int &a, const TensorDim &b) {
+    return a + b.getFeatureLen();
+  };
+
+  auto sample_size =
+    std::accumulate(input_dims.begin(), input_dims.end(), 0u, size_accumulator);
+  sample_size = std::accumulate(label_dims.begin(), label_dims.end(),
+                                sample_size, size_accumulator);
+
+  auto path_prop = std::get<PropsPath>(*raw_file_props);
+  auto file_size = path_prop.file_size();
+
+  /// checking alignment is a good way to make check if a file is valid,
+  /// unfortunately, our test dataset does not have this property
+  /// (trainingSet.dat, valSet.dat, testSet.dat) after checking, we can
+  /// uncomment below line.
+  // NNTR_THROW_IF((file_size % sample_size * RawFileDataProducer::pixel_size !=
+  // 0),
+  //               std::invalid_argument)
+  //   << " Given file does not align with the given sample size, sample size: "
+  //   << sample_size << " file_size: " << file_size;
+
+  return path_prop.file_size() /
+         (sample_size * RawFileDataProducer::pixel_size);
+}
+
+void RawFileDataProducer::setProperty(
+  const std::vector<std::string> &properties) {
+  auto left = loadProperties(properties, *raw_file_props);
+  NNTR_THROW_IF(!left.empty(), std::invalid_argument)
+    << "There is unparsed properties, size: " << left.size();
+}
+
+DataProducer::Generator
+RawFileDataProducer::finalize(const std::vector<TensorDim> &input_dims,
+                              const std::vector<TensorDim> &label_dims) {
+
+  /****************** Validation ****************/
+  auto sz = size(input_dims, label_dims);
+  auto batch = input_dims[0].batch();
+
+  NNTR_THROW_IF(sz < batch, std::invalid_argument)
+    << "calculated sample size is less than a batch";
+
+  auto path_prop = std::get<PropsPath>(*raw_file_props);
+
+  auto size_accumulator = [](const unsigned int &a, const TensorDim &b) {
+    return a + b.getFeatureLen();
+  };
+
+  auto sample_size =
+    std::accumulate(input_dims.begin(), input_dims.end(), 0u, size_accumulator);
+  sample_size = std::accumulate(label_dims.begin(), label_dims.end(),
+                                sample_size, size_accumulator);
+
+  /// below works when checking alignment is correct
+  // auto sample_size = path_prop.file_size() / (sz *
+  // RawFileDataProducer::pixel_size);
+
+  /****************** Prepare states ****************/
+  std::mt19937 rng_;
+  rng_.seed(getSeed());
+  auto idxes_ = std::vector<unsigned int>();
+  idxes_.reserve(sz);
+  /// idxes point to the file position in bytes where a sample starts
+  std::generate_n(std::back_inserter(idxes_), sz,
+                  [sample_size, current = 0ULL]() mutable {
+                    auto c = current;
+                    current += sample_size * RawFileDataProducer::pixel_size;
+                    return c;
+                  });
+  std::shuffle(idxes_.begin(), idxes_.end(), rng_);
```

Comment:
I recommend to remove shuffle from here. Shuffling will be done from the external caller.
If you want to keep it for later remembrance, then consider adding as a comment.

Reply:
Yes, I'll add a comment here

Reply:
Done